File: Java/HashTable/Service/Hashtable.java
package HashTable.Service;

import java.util.Arrays;

public class Hashtable {
    private int[] hashArray;

    public int[] getHashArray() {
        Arrays.stream(this.hashArray).forEach(i -> System.out.println(i));
        return hashArray;
    }

    private int size = 5;
    public Hashtable(int size) {
        if (size!=0){
            this.size = size;
        }
        this.hashArray = new int[this.size];
        initializeArray(this.size);
    }

    private void initializeArray(int size) {
        for (int i = 0; i< this.hashArray.length; i++) {
            this.hashArray[i] =-1;
        }
    }

    private int hash(int key){
        return key % this.size;
    }

    public void addValue( int key, int value){
        if (isFull()){
            System.out.println("Hash Table is full, no more room to enter new values.");
            return;
        }
        int index = hash(key);
        if (this.hashArray[index] == -1){
            this.hashArray[index] = value;
            System.out.println("Value entered successfully");
        }else {
            addValue(index+1, value);
        }
    }

    public boolean isFull() {
        return Arrays.stream(this.hashArray).filter(i -> i!=-1).count() == this.size;
    }

    public void deleteValue(int key, int value){
        if (isEmpty()){
            System.out.println("Hash table is empty.");
            return;
        }
        int index = hash(key);
        if (this.hashArray[index] == value){
            this.hashArray[index] = -1;
        }else{
            deleteValue(index+1, value);
        }
    }

    public boolean isEmpty() { return Arrays.stream(this.hashArray).filter(i -> i!=-1).count() == 0; }

    public void displayAll() {
        if (isEmpty()){
            System.out.println("Hash table is empty.");
            return;
        }
        for (int i: this.hashArray) {
            System.out.println("--->"+i);
        }
    }

    public void peek() {
        if (isEmpty()){
            System.out.println("Hash table is empty.");
            return;
        }
        System.out.println("--->"+this.hashArray[this.size-1]);
    }
}

File: GO-Lang/Hashtable/Services/Hashtable.go
package services

import "fmt"

type (
	HashTable struct {
		size      int
		hashArray []int
	}
)

func InitHashTable(size int) HashTable {
	if size > 0 {
		arr := make([]int, size)
		for i := 0; i < len(arr); i++ {
			arr[i] = -1
		}
		return HashTable{size: size, hashArray: arr}
	}
	return HashTable{size: 5, hashArray: []int{-1, -1, -1, -1, -1}}
}

func (arr *HashTable) IsFull() bool {
	for i, _ := range arr.hashArray {
		if arr.hashArray[i] == -1 {
			return false
		}
	}
	return true
}

func (arr *HashTable) IsEmpty() bool {
	for i, _ := range arr.hashArray {
		if arr.hashArray[i] != -1 {
			return false
		}
	}
	return true
}

func (arr *HashTable) AddValue() {
	if arr.IsFull() {
		fmt.Println("Hash table is full")
		return
	}
	var value int
	fmt.Println("Enter the value you want to enter in the HashTable")
	fmt.Scan(&value)
	index := hash(value, arr.size)

	if arr.hashArray[index] != -1 {
		fmt.Println("index ", index, " is available")
		return
	}
	arr.hashArray[index] = value
}

func hash(value int, size int) int {
	return value % size
}

func (arr *HashTable) Remove(value int) {
	if arr.IsEmpty() {
		fmt.Println("Hash table is Empty")
		return
	}
	index := hash(value, arr.size)
	fmt.Println("Removed value successfully")
	arr.hashArray[index] = -1
}

func (arr *HashTable) DisplayAll() {
	if arr.IsEmpty() {
		fmt.Println("Hash table is Empty")
		return
	}

	for i, v := range arr.hashArray {
		fmt.Println(i, "-->", v)
	}
}

func (arr *HashTable) Peek() {
	if arr.IsEmpty() {
		fmt.Println("Hash table is Empty")
		return
	}

	fmt.Println("---->", arr.hashArray[len(arr.hashArray)-1])
}
